src/config.cpp

```diff
@@ -32,6 +32,15 @@
 
 #include "config_impl.hpp"
 
+namespace {
+
+std::shared_ptr<rapidjson::Document>& get_source() {
```

Comment:
There's no need to return `std::shared_ptr` from the function, `radpidjson::Document &` should be enough.

You could "reset" `rapidjson::Document` with either `operator =(radpijson::Document &&)` or with `Swap(rapidjson::Document &)`:
https://github.com/Tencent/rapidjson/blob/v1.1.0/include/rapidjson/document.h#L2106

The latter is preferable as `operator=` support is controlled by `RAPIDJSON_HAS_CXX11_RVALUE_REFS` define.

Looks like you could get rid of `std::shared_ptr` (as well as usage of singleton pattern) by keeping `source` as `rapidjson::Document`.